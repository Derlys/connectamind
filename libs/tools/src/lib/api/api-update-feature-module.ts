import { Tree } from '@nrwl/devkit'
import { addNamedImport } from '../utils/add-named-import'
import { getDecoratorArgs } from '../utils/get-decorator-args'
import { updateSourceFile } from '../utils/update-source-file'

export function apiUpdateFeatureModule(
  tree: Tree,
  path: string,
  {
    dataAccessClass,
    dataAccessPackage,
    targetClass,
  }: {
    dataAccessClass: string
    dataAccessPackage: string
    targetClass: string
  },
) {
  updateSourceFile(tree, path, (source) => {
    // Add CoreDataAccessModule Import to DataAccessModule
    addNamedImport(source, dataAccessPackage, dataAccessClass)

    // Get the module decorator
    const moduleDecorator = getDecoratorArgs(source, targetClass, 'Module')

    // Remove the 'controllers' property because it's not needed in the feature module
    const hasControllersProperty = moduleDecorator.getPropertyOrThrow('controllers')
    if (hasControllersProperty) {
      hasControllersProperty.remove()
    }

    // Remove the 'exports' property because it's not needed in the feature module
    const hasExportsProperty = moduleDecorator.getPropertyOrThrow('exports')
    if (hasExportsProperty) {
      hasExportsProperty.remove()
    }

    // Add the data-access module to the 'imports' property
    const hasImportsProperty = moduleDecorator.getProperty('imports')
    if (hasImportsProperty) {
      throw new Error(`apiImportModule: ${dataAccessClass} already has an 'imports' property`)
    }
    moduleDecorator.insertPropertyAssignment(0, {
      name: 'imports',
      initializer: `[${dataAccessClass}]`,
    })

    return source
  })
}
