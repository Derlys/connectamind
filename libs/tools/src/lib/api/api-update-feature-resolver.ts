import { Tree } from '@nrwl/devkit'

import { addConstructors } from '../utils/add-constructors'
import { addNamedImport } from '../utils/add-named-import'
import { updateSourceFile } from '../utils/update-source-file'

export function apiUpdateFeatureResolver(
  tree: Tree,
  path: string,
  {
    dataAccessClass,
    dataAccessPackage,
    importProperty,
    targetClass,
  }: {
    dataAccessClass: string
    dataAccessPackage: string
    importProperty: string
    targetClass: string
  },
) {
  updateSourceFile(tree, path, (source) => {
    addNamedImport(source, dataAccessPackage, dataAccessClass)

    addConstructors(source, targetClass, [{ name: importProperty, type: dataAccessClass, private: true }])

    return source
  })
}
