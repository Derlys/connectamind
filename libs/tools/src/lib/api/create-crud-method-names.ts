import { names } from '@nx/devkit'

export function createCrudMethodNames(modelName: string, prefix = '') {
  const model = names(modelName).className

  return {
    createMethod: names(`${prefix ? prefix + '-' : ''}create${model}`).propertyName,
    deleteMethod: names(`${prefix ? prefix + '-' : ''}delete${model}`).propertyName,
    findManyMethod: names(`${prefix ? prefix + '-' : ''}findMany${model}`).propertyName,
    findManyCountMethod: names(`${prefix ? prefix + '-' : ''}findMany${model}Count`).propertyName,
    findOneMethod: names(`${prefix ? prefix + '-' : ''}findOne${model}`).propertyName,
    updateMethod: names(`${prefix ? prefix + '-' : ''}update${model}`).propertyName,
  }
}
