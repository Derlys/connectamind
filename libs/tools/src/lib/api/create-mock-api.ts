import { Tree } from '@nx/devkit'
import { applicationGenerator, libraryGenerator, serviceGenerator } from '@nx/nest'

export async function createMockApi(tree: Tree, app: string) {
  // Build the mock app and core libs
  await applicationGenerator(tree, { name: app, directory: 'apps' })
  // Create the core data access lib
  await libraryGenerator(tree, { name: `data-access`, directory: `libs/${app}/core` })
  // Create the core service
  await serviceGenerator(tree, {
    directory: 'lib',
    flat: true,
    name: `${app}-core`,
    project: `${app}-core-data-access`,
  })
  // Create the core feature lib
  await libraryGenerator(tree, { name: `feature`, directory: `libs/${app}/core` })
}
