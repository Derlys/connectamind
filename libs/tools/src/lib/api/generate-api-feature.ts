import { Tree } from '@nrwl/devkit'
import { NormalizedApiFeatureSchema } from '../../generators/api-feature/api-feature-schema'
import { generateApiLib } from './generate-api-lib'

export async function generateApiFeature(tree: Tree, options: NormalizedApiFeatureSchema) {
  if (!options.skipDataAccess) {
    await generateApiLib(tree, 'data-access', options)
  }
  if (!options.skipFeature) {
    await generateApiLib(tree, 'feature', options)
  }
  if (!options.skipUtil) {
    await generateApiLib(tree, 'util', options)
  }
  if (!options.skipSdk) {
    // throw new Error('generateApiFeature: skipSdk is not implemented')
  }
}
