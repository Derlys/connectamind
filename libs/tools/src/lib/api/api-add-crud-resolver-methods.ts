import { Tree } from '@nrwl/devkit'
import { names } from '@nx/devkit'
import { addNamedImport } from '../utils/add-named-import'
import { createCrudMethodNames } from './create-crud-method-names'
import { getDtoNames } from './get-dto-names'
import { updateSourceFile } from '../utils/update-source-file'

export function apiAddCrudResolverMethods(
  tree: Tree,
  path: string,
  {
    authDataAccessImportPath,
    dataAccessImportPath,
    modelName,
    pluralModelName,
    targetClass,
  }: {
    authDataAccessImportPath: string
    dataAccessImportPath: string
    modelName: string
    pluralModelName: string
    targetClass: string
  },
) {
  const { modelClassName, updateInputClass, createInputClass } = getDtoNames({
    modelName,
  })

  const { findManyMethod, findOneMethod, createMethod, updateMethod, deleteMethod } = createCrudMethodNames(
    modelName,
    pluralModelName,
    'admin',
  )

  const statement = `return this.service`
  const idProperty = names(`${modelName}-id`).propertyName

  updateSourceFile(tree, path, (source) => {
    addNamedImport(source, authDataAccessImportPath, ['ApiAuthGraphqlGuard', 'CtxUser'])
    addNamedImport(source, authDataAccessImportPath.replace('auth/', 'user/'), ['User'])
    addNamedImport(source, '@nestjs/graphql', ['Mutation', 'Query', 'Args'])
    addNamedImport(source, '@nestjs/common', ['UseGuards'])
    addNamedImport(source, dataAccessImportPath, [createInputClass, updateInputClass, modelClassName])

    source.getClass(targetClass).addDecorator({
      name: 'UseGuards',
      arguments: ['ApiAuthGraphqlGuard'],
    })

    source.getClass(targetClass).addMethods([
      {
        decorators: [{ name: 'Mutation', arguments: [`() => ${modelClassName}`, '{ nullable: true }'] }],
        name: createMethod,
        parameters: [ctxUserArgs(), inputArgs(createInputClass)],
        statements: [`${statement}.${createMethod.replace('adminC', 'admin.c')}(user.id, input)`],
      },
      {
        decorators: [{ name: `Mutation`, arguments: [`() => ${modelClassName}`, '{ nullable: true }'] }],
        name: deleteMethod,
        parameters: [ctxUserArgs(), idPropertyArgs(idProperty)],
        statements: [`${statement}.${deleteMethod.replace('adminD', 'admin.d')}(user.id, ${idProperty})`],
      },
      {
        decorators: [{ name: 'Query', arguments: [`() => [${modelClassName}]`, '{ nullable: true }'] }],
        name: findManyMethod,
        parameters: [ctxUserArgs()],
        statements: [`${statement}.${findManyMethod.replace('adminF', 'admin.f')}(user.id)`],
      },
      {
        decorators: [{ name: 'Query', arguments: [`() => ${modelClassName}`, '{ nullable: true }'] }],
        name: findOneMethod,
        parameters: [ctxUserArgs(), idPropertyArgs(idProperty)],
        statements: [`${statement}.${findOneMethod.replace('adminF', 'admin.f')}(user.id, ${idProperty})`],
      },
      {
        decorators: [{ name: 'Mutation', arguments: [`() => ${modelClassName}`, '{ nullable: true }'] }],
        name: updateMethod,
        parameters: [ctxUserArgs(), idPropertyArgs(idProperty), inputArgs(updateInputClass)],
        statements: [`${statement}.${updateMethod.replace('adminU', 'admin.u')}(user.id, ${idProperty}, input)`],
      },
    ])

    return source
  })
}

function idPropertyArgs(idProperty: string) {
  return {
    name: `${idProperty}`,
    type: 'string',
    decorators: [{ name: 'Args', arguments: [`'${idProperty}'`] }],
  }
}

function inputArgs(inputClass: string) {
  return {
    name: 'input',
    type: inputClass,
    decorators: [{ name: 'Args', arguments: [`'input'`] }],
  }
}

function ctxUserArgs() {
  return {
    name: 'user',
    type: 'User',
    decorators: [{ name: 'CtxUser()' }],
  }
}
