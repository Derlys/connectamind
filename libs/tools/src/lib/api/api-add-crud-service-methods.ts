import { Tree } from '@nrwl/devkit'
import { names } from '@nx/devkit'
import { dirname } from 'node:path'
import { addExport } from '../utils/add-export'
import { addNamedImport } from '../utils/add-named-import'
import { updateSourceFile } from '../utils/update-source-file'
import { createCrudMethodNames } from './create-crud-method-names'
import { getDtoNames } from './get-dto-names'
import { getFindManyHelper } from './get-find-many-helper'
import { getInputTypeCreateUpdate } from './get-input-type-create-update'
import { getInputTypeFindMany } from './get-input-type-find-many'
import { getObjectTypeContent } from './get-object-type-content'

export function apiAddCrudServiceMethods(
  tree: Tree,
  path: string,
  {
    modelName,
    pluralModelName,
    label,
    targetClass,
  }: { modelName: string; pluralModelName: string; label: string; targetClass: string },
) {
  const {
    createInputClass,
    createInputFile,
    findManyHelperFile,
    findManyInputClass,
    findManyInputFile,
    modelClassName,
    updateInputClass,
    updateInputFile,
  } = getDtoNames({
    modelName,
  })

  const { createMethod, deleteMethod, findManyCountMethod, findManyMethod, findOneMethod, updateMethod } =
    createCrudMethodNames(modelName, pluralModelName)

  const statement = `this.core.data.${names(modelName).propertyName}`
  const idProperty = names(`${modelName}-id`).propertyName

  // Create the DTO's for the create and update methods
  const basePath = dirname(path)
  const dtoPath = `${basePath}/dto`
  const helpersPath = `${basePath}/helpers`

  const createInputDtoPath = `${dtoPath}/${names(createInputFile).fileName}.ts`
  const findManyHelper = `${helpersPath}/${names(findManyHelperFile).fileName}.ts`
  const findManyInputDtoPath = `${dtoPath}/${names(findManyInputFile).fileName}.ts`
  const updateInputDtoPath = `${dtoPath}/${names(updateInputFile).fileName}.ts`

  tree.write(createInputDtoPath, getInputTypeCreateUpdate(createInputClass, `${label}!`))
  tree.write(findManyHelper, getFindManyHelper(findManyHelperFile, findManyInputClass, findManyInputFile, modelName))
  tree.write(findManyInputDtoPath, getInputTypeFindMany(findManyInputClass))
  tree.write(updateInputDtoPath, getInputTypeCreateUpdate(updateInputClass, `${label}?`))

  // Create the entity file
  const entityFile = `${names(modelName).fileName}.entity.ts`
  const entityPath = `${dirname(path)}/entity/${entityFile}`
  tree.write(entityPath, getObjectTypeContent(modelClassName, label))

  const { propertyName: parsePropertyName } = names(findManyHelperFile)

  // Add the exports to the barrel file
  const barrelPath = `${dirname(path)}/../index.ts`
  addExport(tree, barrelPath, `./lib/dto/${names(createInputFile).fileName}`)
  addExport(tree, barrelPath, `./lib/dto/${names(findManyInputFile).fileName}`)
  addExport(tree, barrelPath, `./lib/dto/${names(updateInputFile).fileName}`)
  addExport(tree, barrelPath, `./lib/entity/${entityFile}`)

  updateSourceFile(tree, path, (source) => {
    addNamedImport(source, `./dto/${names(createInputFile).fileName}`, createInputClass)
    addNamedImport(source, `./dto/${names(findManyInputFile).fileName}`, findManyInputClass)
    addNamedImport(source, `./dto/${names(updateInputFile).fileName}`, updateInputClass)
    addNamedImport(source, `./helpers/${names(findManyHelperFile).fileName}`, parsePropertyName)

    source.getClass(targetClass).addMethods([
      {
        isAsync: true,
        name: createMethod,
        parameters: [
          { name: 'adminId', type: 'string' },
          { name: 'input', type: createInputClass },
        ],
        statements: ['await this.core.ensureUserAdmin(adminId)', `return ${statement}.create({ data: input })`],
      },
      {
        isAsync: true,
        name: deleteMethod,
        parameters: [
          { name: 'adminId', type: 'string' },
          { name: `${idProperty}`, type: 'string' },
        ],
        statements: [
          'await this.core.ensureUserAdmin(adminId)',
          `const deleted = await ${statement}.delete({ where: { id: ${idProperty} }})`,
          `return !!deleted`,
        ],
      },
      {
        isAsync: true,
        name: findManyMethod,
        parameters: [
          { name: 'adminId', type: 'string' },
          { name: 'input', type: findManyInputClass },
        ],
        statements: [
          'await this.core.ensureUserAdmin(adminId)',
          `const { where, orderBy, take, skip } = ${parsePropertyName}(input)`,
          `const items = await ${statement}.findMany({ where, orderBy, take, skip })`,
          `return items ?? []`,
        ],
      },
      {
        isAsync: true,
        name: findManyCountMethod,
        parameters: [
          { name: 'adminId', type: 'string' },
          { name: 'input', type: findManyInputClass },
        ],
        statements: [
          'await this.core.ensureUserAdmin(adminId)',
          `const { where, orderBy, take, skip } = ${parsePropertyName}(input)`,
          `const [count, total] = await Promise.all([${statement}.count({ where, orderBy, take, skip }), ${statement}.count({ where, orderBy }), ])`,
          `return { count, skip, take, total }`,
        ],
      },
      {
        isAsync: true,
        name: findOneMethod,
        parameters: [
          { name: 'adminId', type: 'string' },
          { name: `${idProperty}`, type: 'string' },
        ],
        statements: [
          'await this.core.ensureUserAdmin(adminId)',
          `return ${statement}.findUnique({ where: { id: ${idProperty} } })`,
        ],
      },
      {
        isAsync: true,
        name: updateMethod,
        parameters: [
          { name: 'adminId', type: 'string' },
          { name: `${idProperty}`, type: 'string' },
          { name: 'input', type: updateInputClass },
        ],
        statements: [
          'await this.core.ensureUserAdmin(adminId)',
          `return ${statement}.update({ where: { id: ${idProperty} }, data: input })`,
        ],
      },
    ])

    return source
  })
}
