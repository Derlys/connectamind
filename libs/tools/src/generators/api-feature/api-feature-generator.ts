import { formatFiles, Tree } from '@nx/devkit'
import { generateApiFeature } from '../../lib/api/generate-api-feature'
import { normalizeApiFeatureSchema } from '../../lib/api/normalize-api-feature-schema'
import { ensureAppExists } from '../../lib/utils/ensure-app-exists'
import { ApiFeatureGeneratorSchema } from './api-feature-schema'

export async function apiFeatureGenerator(tree: Tree, rawOptions: ApiFeatureGeneratorSchema) {
  const options = normalizeApiFeatureSchema(rawOptions)
  ensureAppExists(tree, options.app)
  await generateApiFeature(tree, options)
  await formatFiles(tree)
}

export default apiFeatureGenerator
