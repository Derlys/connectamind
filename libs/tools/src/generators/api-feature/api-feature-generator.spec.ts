import { readProjectConfiguration, Tree } from '@nx/devkit'
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing'
import { createMockApi } from '../../lib/api/create-mock-api'

import { apiFeatureGenerator } from './api-feature-generator'
import { ApiFeatureGeneratorSchema } from './api-feature-schema'

describe('api-feature generator', () => {
  let tree: Tree
  const options: ApiFeatureGeneratorSchema = { app: 'api', name: 'test', label: 'name' }

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace()
  })

  it('should generate the feature libraries', async () => {
    await createMockApi(tree, options.app)

    // By default, we generate two libraries: data-access and feature
    const libs = ['data-access', 'feature']
    await apiFeatureGenerator(tree, options)

    libs.forEach((lib) => {
      const config = readProjectConfiguration(tree, `${options.app}-${options.name}-${lib}`)
      expect(config).toBeDefined()
    })

    const basePathDataAccess = `libs/${options.app}/${options.name}/data-access/src`
    const dataAccessBarrel = `${basePathDataAccess}/index.ts`
    expect(tree.exists(dataAccessBarrel)).toBeTruthy()
    expect(tree.read(dataAccessBarrel).toString()).toMatchSnapshot()

    const dataAccessService = `${basePathDataAccess}/lib/${options.app}-${options.name}.service.ts`
    expect(tree.exists(dataAccessService)).toBeTruthy()
    expect(tree.read(dataAccessService).toString()).toMatchSnapshot()

    const dataAccessAdminService = `${basePathDataAccess}/lib/${options.app}-${options.name}-admin.service.ts`
    expect(tree.exists(dataAccessAdminService)).toBeTruthy()
    expect(tree.read(dataAccessAdminService).toString()).toMatchSnapshot()

    const basePathFeature = `libs/${options.app}/${options.name}/feature/src`
    const featureBarrel = `${basePathFeature}/index.ts`
    expect(tree.exists(featureBarrel)).toBeTruthy()
    expect(tree.read(featureBarrel).toString()).toMatchSnapshot()

    const dataAccessResolver = `${basePathFeature}/lib/${options.app}-${options.name}.resolver.ts`
    expect(tree.exists(dataAccessResolver)).toBeTruthy()
    expect(tree.read(dataAccessResolver).toString()).toMatchSnapshot()

    const dataAccessAdminResolver = `${basePathFeature}/lib/${options.app}-${options.name}-admin.resolver.ts`
    expect(tree.exists(dataAccessAdminResolver)).toBeTruthy()
    expect(tree.read(dataAccessAdminResolver).toString()).toMatchSnapshot()
  })

  it('should generate the feature libraries with util lib', async () => {
    await createMockApi(tree, options.app)

    // By default, we generate two libraries: data-access and feature
    const libs = ['data-access', 'feature', 'util']
    await apiFeatureGenerator(tree, { ...options, skipUtil: false })

    libs.forEach((lib) => {
      const config = readProjectConfiguration(tree, `${options.app}-${options.name}-${lib}`)
      expect(config).toBeDefined()
    })
  })

  it('should generate the feature with different name', async () => {
    const testOptions = { ...options, name: 'company' }
    await createMockApi(tree, testOptions.app)

    // By default, we generate two libraries: data-access and feature
    const libs = ['data-access', 'feature', 'util']
    await apiFeatureGenerator(tree, { ...testOptions, skipUtil: false })

    libs.forEach((lib) => {
      const config = readProjectConfiguration(tree, `${testOptions.app}-${testOptions.name}-${lib}`)
      expect(config).toBeDefined()
    })
  })

  it('should generate the feature libraries with custom names', async () => {
    await createMockApi(tree, options.app)

    // By default, we generate two libraries: data-access and feature
    const libs = ['data-access', 'feature']
    const customName = 'custom'
    await apiFeatureGenerator(tree, { ...options, name: customName })

    libs.forEach((lib) => {
      const config = readProjectConfiguration(tree, `${options.app}-${customName}-${lib}`)
      expect(config).toBeDefined()
    })
  })
})
