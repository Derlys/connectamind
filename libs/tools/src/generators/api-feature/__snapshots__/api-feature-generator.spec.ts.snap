// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-feature generator should generate the feature libraries 1`] = `
"export * from './lib/api-test-data-access.module';
export * from './lib/api-test.service';
export * from './lib/entity/test.entity';
export * from './lib/dto/admin-create-test.input';
export * from './lib/dto/admin-find-many-test.input';
export * from './lib/dto/admin-update-test.input';
"
`;

exports[`api-feature generator should generate the feature libraries 2`] = `
"import { Injectable } from '@nestjs/common';
import { ApiCoreService } from '@proj/api/core/data-access';
import { AdminCreateTestInput } from './dto/admin-create-test.input';
import { AdminFindManyTestInput } from './dto/admin-find-many-test.input';
import { AdminUpdateTestInput } from './dto/admin-update-test.input';
import { parseAdminFindManyTest } from './helpers/parse-admin-find-many-test';

@Injectable()
export class ApiTestAdminService {
  constructor(private readonly core: ApiCoreService) {}

  async createTest(input: AdminCreateTestInput) {
    return this.core.data.test.create({ data: input });
  }

  async deleteTest(testId: string) {
    const deleted = await this.core.data.test.delete({ where: { id: testId } });
    return !!deleted;
  }

  async findManyTest(input: AdminFindManyTestInput) {
    const { where, orderBy, take, skip } = parseAdminFindManyTest(input);
    const items = await this.core.data.test.findMany({
      where,
      orderBy,
      take,
      skip,
    });
    return items ?? [];
  }

  async findManyTestCount(input: AdminFindManyTestInput) {
    const { where, orderBy, take, skip } = parseAdminFindManyTest(input);
    const [count, total] = await Promise.all([
      this.core.data.test.count({ where, orderBy, take, skip }),
      this.core.data.test.count({ where, orderBy }),
    ]);
    return { count, skip, take, total };
  }

  async findOneTest(testId: string) {
    return this.core.data.test.findUnique({ where: { id: testId } });
  }

  async updateTest(testId: string, input: AdminUpdateTestInput) {
    return this.core.data.test.update({ where: { id: testId }, data: input });
  }
}
"
`;

exports[`api-feature generator should generate the feature libraries 3`] = `
"import { Module } from '@nestjs/common';
import { ApiCoreDataAccessModule } from '@proj/api/core/data-access';
import { ApiTestAdminService } from './api-test-admin.service';

import { ApiTestService } from './api-test.service';

@Module({
  imports: [ApiCoreDataAccessModule],
  providers: [ApiTestService, ApiTestAdminService],
  exports: [ApiTestService],
})
export class ApiTestDataAccessModule {}
"
`;

exports[`api-feature generator should generate the feature libraries 4`] = `
"import { Injectable } from '@nestjs/common';
import { ApiCoreService } from '@proj/api/core/data-access';
import { ApiTestAdminService } from './api-test-admin.service';

@Injectable()
export class ApiTestService {
  constructor(
    private readonly core: ApiCoreService,
    readonly admin: ApiTestAdminService
  ) {}
}
"
`;

exports[`api-feature generator should generate the feature libraries 5`] = `
"import { Field, InputType } from '@nestjs/graphql';

@InputType()
export class AdminCreateTestInput {
  @Field()
  name!: string;
}
"
`;

exports[`api-feature generator should generate the feature libraries 6`] = `
"import { Field, InputType, Int } from '@nestjs/graphql';

@InputType()
export class AdminFindManyTestInput {
  @Field({ nullable: true })
  search?: string;
  @Field(() => Int, { nullable: true, defaultValue: 0 })
  skip?: number;
  @Field(() => Int, { nullable: true, defaultValue: 10 })
  take?: number;
}
"
`;

exports[`api-feature generator should generate the feature libraries 7`] = `
"import { Field, InputType } from '@nestjs/graphql';

@InputType()
export class AdminUpdateTestInput {
  @Field({ nullable: true })
  name?: string;
}
"
`;

exports[`api-feature generator should generate the feature libraries 8`] = `
"import { Prisma } from '@prisma/client';
import { AdminFindManyTestInput } from '../dto/admin-find-many-test.input';

export function parseAdminFindManyTest(input: AdminFindManyTestInput): {
  orderBy: Prisma.TestOrderByWithRelationInput;
  skip?: number;
  take?: number;
  where: Prisma.TestWhereInput;
} {
  const where: Prisma.TestWhereInput = {};

  if (input.search) {
    where.OR = [
      { id: { contains: input.search, mode: 'insensitive' } },
      { name: { contains: input.search, mode: 'insensitive' } },
    ];
  }

  return {
    where,
    skip: input.skip ?? 0,
    take: input.take ?? 10,
    orderBy: { name: 'asc' },
  };
}
"
`;

exports[`api-feature generator should generate the feature libraries 9`] = `
"import { Field, ObjectType } from '@nestjs/graphql';

@ObjectType()
export class Test {
  @Field()
  id!: string;
  @Field({ nullable: true })
  createdAt?: Date;
  @Field({ nullable: true })
  updatedAt?: Date;
  @Field()
  name!: string;
}
"
`;

exports[`api-feature generator should generate the feature libraries 10`] = `
"export * from './lib/api-test-feature.module';
"
`;

exports[`api-feature generator should generate the feature libraries 11`] = `
"import { Resolver } from '@nestjs/graphql';
import { ApiTestService } from '@proj/api/test/data-access';
import { ApiAuthGraphQLAdminGuard } from '@proj/api/auth/data-access';
import { Paging } from '@proj/api/core/data-access';
import { Mutation, Query, Args } from '@nestjs/graphql';
import { UseGuards } from '@nestjs/common';
import {
  AdminCreateTestInput,
  AdminFindManyTestInput,
  Test,
  AdminUpdateTestInput,
} from '@proj/api/test/data-access';

@Resolver()
@UseGuards(ApiAuthGraphQLAdminGuard)
export class ApiTestAdminResolver {
  constructor(private readonly service: ApiTestService) {}

  @Mutation(() => Test, { nullable: true })
  adminCreateTest(@Args('input') input: AdminCreateTestInput) {
    return this.service.admin.createTest(input);
  }

  @Mutation(() => Boolean, { nullable: true })
  adminDeleteTest(@Args('testId') testId: string) {
    return this.service.admin.deleteTest(testId);
  }

  @Query(() => [Test], { nullable: true })
  adminFindManyTest(@Args('input') input: AdminFindManyTestInput) {
    return this.service.admin.findManyTest(input);
  }

  @Query(() => Paging, { nullable: true })
  adminFindManyTestCount(@Args('input') input: AdminFindManyTestInput) {
    return this.service.admin.findManyTestCount(input);
  }

  @Query(() => Test, { nullable: true })
  adminFindOneTest(@Args('testId') testId: string) {
    return this.service.admin.findOneTest(testId);
  }

  @Mutation(() => Test, { nullable: true })
  adminUpdateTest(
    @Args('testId') testId: string,
    @Args('input') input: AdminUpdateTestInput
  ) {
    return this.service.admin.updateTest(testId, input);
  }
}
"
`;

exports[`api-feature generator should generate the feature libraries 12`] = `
"import { Module } from '@nestjs/common';
import { ApiTestDataAccessModule } from '@proj/api/test/data-access';
import { ApiTestAdminResolver } from './api-test-admin.resolver';

@Module({
  imports: [ApiTestDataAccessModule],
  providers: [ApiTestAdminResolver],
})
export class ApiTestFeatureModule {}
"
`;
